Add support for disabling signatures

crimsonknave opened the pull request:
I'd like to be able to disable signatures so that I can filter down signatures that I either want to deal with later or don't find useful. This is especially when trying to roll out gitrob across a large organization. 

This PR isn't final, I wanted to hammer out a basic outline and then poll for thoughts. (It does work though)
- I'm not sure if I like having just an array of patterns.
  - I didn't notice any duplicate patterns that were match vs regex, but it could happen I believe
  - However, it seemed cumbersome to basically duplicate the signature just to disable it
  - Ideally, it would be an array of names/ids for the signature, but there isn't such a field currently.
- Gotta write tests, but the tests don't pass for me on master. I suspect I'm just missing some config, but there wasn't much documentation in the readme

michenriksen commented:
This looks very interesting! It's cool to have the possibility to tune the signatures.

I was wondering if it would make sense to define the ignore-signatures a bit like the actual signatures, something like this:

    [
        {
            "part": "extension",
            "type": "match",
            "pattern": "log"
        },
        {
            "part": "path",
            "type": "regex",
            "pattern": "/test/"
        }
    ] 

The above signatures would instruct Gitrob to ignore any files with the `.log` extension as well as any file where the word `test` is in the complete path.

I think this would make it very flexible and it's nice that it follows a familiar "syntax".

What do you think?

crimsonknave commented:
Yeah, I was wavering back and forth on how I wanted to do it. I opted for just the pattern for simplicity to start. 

There are a couple of ways I see that this could be implemented:
* More or less how I have it. It's pretty simple, but also more limited.
* You can disable existing signatures by specifying the part, type and pattern.
* You can specify signatures of files to *ignore* by specifying part, type and pattern. So, instead of just working with existing signatures you can say never flag a file that matches this pattern.
* Making the signatures a bit more robust by adding a state and an id. Additionally the signatures would be stored in a hash with the id as a key. Then the custom signatures could update the base signatures and set the disabled state by just including the id and the state.

I like the third and fourth the best, but they are much more extensive changes. It might also make sense to implement both the third and the fourth as they server different features. By ignoring the exact specification of a server you get the effect of disabling it, but if the signatures had an id and that was displayed in the ui, then it would be much easier to know how to disable one if you could just use that instead of digging into the source code to figure out which one matched the file.

The fourth one alters the  format of the signatures file and would either need some finessing to get the old format to work as well or would be a breaking change. If we go with the fourth one it might also make sense to add a category to the signatures since we'd be altering the format anyway, that way down the road you could set something (in your .gitrobrc for example) that says I only want to look at signatures with the 'key' category.

Just throwing thoughts out there, let me know what you think.

crimsonknave commented:
Any thoughts? I'm happy to alter/fix it to whatever makes the most sense.

michenriksen commented:
I'm not too fond of changing the structure of the signatures right now as it would be a breaking change, but it's definitely something to look into in the future. If we implemented a `.gitrobignore` file with signatures to ignore, Gitrob would just have to load those the same way as it loads the other signatures and first run the files through those and skip the file if it matches any ignore signatures. 

This would allow users to define very broad ignore patterns e.g. "Ignore all log files regardless of why they were flagged" as well as more precise ignoring of specific signatures by simply copy-pasting the signature's pattern. 

This could also work as a way to permanently whitelist files if you are certain that it will always be a false-positive. For example an organization that has `credentials.enc` files in their projects, but they are always encrypted and therefore safe. An auditor could then add an ignore pattern that would match this type of file to get rid of them in the assessments.

crimsonknave commented:
Updated to the ignore style. ```.gitrobignore``` will ignore any entries that match. I removed the requirement for caption and description as I wasn't sure how they'd fit into the UI.

That more or less what you were thinking? If so I'll clean it up with some test and so on.

michenriksen commented:
This is excellent, thanks so much for contributing this back, and sorry for being slow. I will have a test run and merge it in. :clap:

crimsonknave commented:
I'm happy to add some tests and rebase down to a single commit. However, the last time I tried to run the tests on master they weren't passing for me.
